execution/gethexec/blockmetadata.go

```diff
@@ -0,0 +1,67 @@
+package gethexec
+
+import (
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/util/containers"
+)
+
+type BlockMetadataFetcher interface {
+	BlockMetadataAtCount(count arbutil.MessageIndex) (arbostypes.BlockMetadata, error)
+	BlockNumberToMessageIndex(blockNum uint64) (arbutil.MessageIndex, error)
+	MessageIndexToBlockNumber(messageNum arbutil.MessageIndex) uint64
+}
+
+type BulkBlockMetadataFetcher struct {
+	fetcher BlockMetadataFetcher
+	cache   *containers.LruCache[arbutil.MessageIndex, arbostypes.BlockMetadata]
+}
+
+func NewBulkBlockMetadataFetcher(fetcher BlockMetadataFetcher, cacheSize int) *BulkBlockMetadataFetcher {
+	var cache *containers.LruCache[arbutil.MessageIndex, arbostypes.BlockMetadata]
+	if cacheSize != 0 {
+		cache = containers.NewLruCache[arbutil.MessageIndex, arbostypes.BlockMetadata](cacheSize)
+	}
+	return &BulkBlockMetadataFetcher{
+		fetcher: fetcher,
+		cache:   cache,
+	}
+}
+
+func (b *BulkBlockMetadataFetcher) Fetch(fromBlock, toBlock hexutil.Uint64) ([]NumberAndBlockMetadata, error) {
```

Comment:
I didn't prefer using rpc.BlockNumber because then we would need to worry about converting string inputs such as "latest" to block numbers. Now that I think about it, maybe its worth it. I'll change the impl to take in rpc.BlockNumber instead of uint64